torch/csrc/dynamo/guards.cpp

```diff
@@ -574,6 +574,480 @@ static struct PyModuleDef _module = {
+ * if they do not alias.
+ */
+class NoTensorAliasingGuard : public LeafGuard {
+ public:
+  NoTensorAliasingGuard() : _is_first_call(true) {}
+
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    if (_is_first_call) {
+      _first_tensor = value;
+      _is_first_call = false;
+      return true;
+    }
+    _is_first_call = true;
+    return _first_tensor != value;
+  }
+
+  std::string get_failure_reason(PyObject* value) override { // borrowed ref
+    return "";
+  }
+
+ private:
+  bool _is_first_call;
+  PyObject* _first_tensor;
+};
+
+class GuardManager;
+/**
+ * Base class representing a pair of accessor and the associated guard manager.
+ * The accessor defines how to access the child value from the py::object given
+ * to the parent check function.
+ *
+ * GuardAccessors can be considered equivalent to name() method of Source
+ * objects in guards.py. In python, name() method returns a str which we can
+ * then eval in f_locals and f_globals to retrieve the actual py object.
+ * GuardAccessor serves the same purpose. The minor difference is that
+ * GuardManager is a tree structure, so a GuardAccessor just has to retrieve the
+ * value in the next level in this tree and pass it to the child GuardAccessor.
+ *
+ * GuardAccessor also owns the GuardManager associated with the retrieved value
+ * from the GuardAccessor.
+ */
+class GuardAccessor {
+ public:
+  GuardAccessor(py::object accessor_key)
+      : _guard_manager(std::make_unique<GuardManager>()),
+        _accessor_key(accessor_key) {}
+
+  // Return by reference as GuardAccessor owns the GuardManager.
+  std::unique_ptr<GuardManager>& get_guard_manager() {
+    return _guard_manager;
+  }
+
+  bool matches_key(const py::object key) const {
+    return _accessor_key.equal(key);
+  }
+
+  virtual ~GuardAccessor() {
+    // _accessed_value is owned by the GuardAccessor. method access returns an
+    // owning reference.
+    Py_DECREF(_accessed_value);
+  }
+
+  // Returns a borrowed reference. The accessed_value is owned by the
+  // GuardAccessor itself and is decref'd on destruction.
+  virtual PyObject* access(PyObject* obj) = 0;
+
+ private:
+  // Guard manager corresponding to the retrieved value from the GuardAccessor.
+  std::unique_ptr<GuardManager> _guard_manager;
+
+ protected:
+  // accessor key could be py::str for getattr, getitem or py::function for
+  // lambda accessor.
+  py::object _accessor_key;
+  // Keeps the pointer to the accessed value to implement RAII mechanism.
+  PyObject* _accessed_value;
+};
+
+/**
+ * Represents __getattr__ acccessor.
+ */
+class GetAttrGuardAccessor : public GuardAccessor {
+ public:
+  GetAttrGuardAccessor(py::str name)
+      : GuardAccessor(name), _attr_name(name.ptr()) {}
+
+  PyObject* access(PyObject* obj) override { // borrowed ref
+    PyObject* ret = PyObject_GetAttr(obj, _attr_name);
+    // Store the owning reference to _accessed_value to be decref'd on
+    // destruction.
+    _accessed_value = ret;
```

Comment:
Ownership model seems wrong here.
1) This object won't get destroyed after guards run.
2) If _accessed_value already has a value in it, that value will get leaked.

Reply:
The value is decref'd in the destructor of GuardAccessor. 
~~~
virtual ~GuardAccessor() {
    // _accessed_value is owned by the GuardAccessor. method access returns an
    // owning reference.
    Py_DECREF(_accessed_value);
  } 
~~~

The idea is that when we walk through the accessors in the tree, the GuardAccssor object going out of scope will decrement the ref count of the _accessed_value (in the following loop).

~~~
    bool failed_on_first = true;
    for (const auto& accessor : _accessors) {
      auto& manager = accessor->get_guard_manager();
      result = result && manager->check_nopybind(accessor->access(value));
      if (!result) {
        break;
      }
      failed_on_first = false;
    }
~~~

I am happy to discuss offline if there is a better way to represent this.

Reply:
`const auto& accessor` means you take a *reference* to the accessor, so it will not be destroyed when it goes out of scope.

You also don't want to be creating/destroying the guard accessor each time, because it holds ownership of the keys (that you want to persist).

I think it might be cleaner to make the sub-guards a property of the accessor, and have access just use a `check()` method that returns a bool.  That way ownership doesn't need to cross function boundaries.

Reply:
Thanks for catching this. It was silly on my end. 

I liked your idea about running subguards from accessor itself. I pursued this and found that calling guard_manager->check from inside of GuardAccessor will create a circular dependency that cannot be met by forward declaration in the cpp file (we already forward declare GuardManger in the cpp file today). I could move things to header file etc, but not sure if this is what we want either.

Other alternative is to keep the current structure same and `access` always return a new reference. So, the Guardmanager code will look something like this

~~~
    bool failed_on_first = true;
    for (const auto& accessor : _accessors) {
      auto& manager = accessor->get_guard_manager();
      PyObject* x  = accessor->access(value); // new ref
      result = result && manager->check_nopybind(x);
      Py_DECREF(x);
      if (!result) {
        break;
      }
      failed_on_first = false;
    }

~~~

The pro here is all code is in one file so simpler to read. Con is that we will have to unnecessary increment a ref count for PyDict_GetItem accessor because that one does not return a borrowed ref. We could make helper functions - `maybe_decref` to avoid that as well. But then its a vtable function call, and just inc/dec the ref count should be faster (not slower) than the vtable call.

Let me know if what are your opinions on this.

Reply:
Circular deps should not be an issue.  Just forward declare in the same file.  Headers versus cpp files doesn't matter in C++, they are all just one file after the `#include`'s are processed.  So anything that can go in a header can go in a cpp.

Reply:
Ok, it works. I just had to move the derived classes of GuardAccessor to the end.

Thanks for the suggestion. The ownership does not cross function boundaries now. Its cleaner now.